### PR TITLE
support params groups, test=develop

### PR DESCRIPTION
**Pull request by jerrywgz**
### PR types

New features

### PR changes

Others

### Describe

Support parameter groups for optimizers and groups could specific different optimization options

Take SGD as example,  parameters receive list of dict as input and each dict must contain params and ignore entry （weight_decay for first group）will take input args (weight_decay=0.01) as default. 
```python
adam = paddle.optimizer.SGD(learning_rate=0.01,
                                    parameters=[{
                                        'params': linear_1.parameters()
                                    }, {
                                        'params': linear_2.parameters(),
                                        'weight_decay': 0.001,
                                        'learning_rate': 0.1
                                    }],
                                    weight_decay=0.01)
```

Note:

1. learning_rate set in parameter groups represents scale of base learning_rate, for example, the actual learning_rate of the second groups is 0.001.
2. If parameter groups are used, learning_rate, decay and grad_clip have higher priority than param attr.

### CODE REVIEW

**Review by jzhang533 — COMMENTED** [review]
- 对动态图下的分布式的功能不确定是否有影响，需要也请分布式的同学review一下。
- 我们是有动转静后训练的功能的，不确定这类对优化器的改动，对动转静的影响，需要也请做动转静的同学review一下。
- paddle.save 和 paddle.load是可以正常保存和加载parameter group的吧。
- 现设置学习率，可以在`ParamAttr`, `optimizer`里的`learning_rate`参数（可以是float，也可以是LRScheduler）里，`parameter group`里设置；相应的，weight_decay也类似，另外，现在paddle也有weigt average(`paddle.incubate.optimizer.ModelAverage`)，我们急需一篇类似于 [torch.optim](https://pytorch.org/docs/stable/optim.html)  这么组织的文档；不然用的人真的会迷茫~

File: python/paddle/optimizer/optimizer.py [hunk 1]
```diff
@@ -91,6 +95,29 @@ class Optimizer(object):
             adam.step()
             adam.clear_grad()
 
+            #Take the subclass sgd as an example
+            #optimize parameters in linear_1 and linear2 in different options. 
+            #Note that the learning_rate of linear_2 is 0.01.
```
Comment (zhiqiu) [inline python/paddle/optimizer/optimizer.py:100]: the learning_rate of linear_2 is 0.01 ?

File: python/paddle/optimizer/optimizer.py [hunk 2]
```diff
@@ -120,14 +148,17 @@ def __init__(self,
...
+                if not isinstance(self._parameter_list[0], dict):
+                    for param in self._parameter_list:
+                        if hasattr(
+                                param,
+                                'regularizer') and param.regularizer is not None:
+                            logging.info(
+                                "If regularizer of a Parameter has been set by 'paddle.ParamAttr' or 'static.WeightNormParamAttr' already. "
+                                "The weight_decay[%s] in Optimizer will not take effect, and it will only be applied to other Parameters!"
+                                % weight_decay.__str__())
+                            break
```
Comment (zhiqiu) [inline python/paddle/optimizer/optimizer.py:165]: For param_group, the priority of weight_decay is param.weight_decay > param_group.weight_decay > optimizer.weight_decay?

File: python/paddle/optimizer/optimizer.py [hunk 3]
```diff
@@ -934,13 +1026,82 @@ def step(self):
...
+    def _add_param_group(self, param_group):
+        """
+        Add a param group to parameter_list.
+
+        Args:
+            param_group (dict): The group of Tensors to be optimzed with
+            different optimization options.
+        """
+        params = param_group['params']
+        if type(params) == Parameter:
```
Comment (zhiqiu) [inline python/paddle/optimizer/optimizer.py]: Better use isinstance

File: python/paddle/optimizer/optimizer.py [hunk 4]
```diff
@@ -934,13 +1026,82 @@ def step(self):
...
+        self.param_groups.append(param_group)
+
+    def _update_param_group(self, parameters):
+        """
+        Update the param group with new entry
+        Args:
+            parameters (dict): The extra group of Tensors to be optimzed with
+            different optimization options.
+        """
+        return parameters
```
Comment (zhiqiu) [inline python/paddle/optimizer/optimizer.py]: is this function complete?

File: python/paddle/optimizer/rmsprop.py [hunk 5]
```diff
@@ -117,6 +120,26 @@ class RMSProp(Optimizer):
             rmsprop.step()
             rmsprop.clear_grad()
 
+            #Note that the learning_rate of linear_2 is 0.01.
```
Comment (zhiqiu) [inline python/paddle/optimizer/rmsprop.py:123]: same before

File: python/paddle/optimizer/rmsprop.py [hunk 6]
```diff
@@ -160,11 +183,20 @@ def __init__(self,
         self._epsilon = epsilon
         self._momentum = momentum
         self._centered = centered
+        self.default_dict = {
+            'rho': rho,
+            'epsilon': epsilon,
+            'momentum': momentum,
+            'centered': centered,
+        }
```
Comment (zhiqiu) [inline python/paddle/optimizer/rmsprop.py]: the default dict will override those in Base class?

File: python/paddle/optimizer/optimizer.py [hunk 7]
```diff
@@ -163,6 +200,18 @@ def __init__(self,
         self._accumulators_holder = {}
         self._param_device_map = dict()
         self.clear_gradients = self.clear_grad
+        self.default_dict = {
```
Comment (chenwhql) [inline python/paddle/optimizer/optimizer.py]: `default_dict` -> `_default_dict`?

File: python/paddle/optimizer/optimizer.py [hunk 8]
```diff
@@ -163,6 +200,18 @@ def __init__(self,
...
+            'learning_rate': self._learning_rate,
+            'weight_decay': self.regularization,
+            'grad_clip': self._grad_clip
+        }
+
+        self.param_groups = []
```
Comment (chenwhql) [inline python/paddle/optimizer/optimizer.py]: same above

File: python/paddle/optimizer/optimizer.py [hunk 9]
```diff
@@ -148,7 +179,13 @@ def __init__(self,
         self._dtype = None
         # Infer the dtype form parameter
         if self._parameter_list:
-            self._dtype = self._parameter_list[0].dtype
+            if isinstance(self._parameter_list[0], dict):
```
Comment (chenwhql) [inline python/paddle/optimizer/optimizer.py:187]: do we need check whether `self._parameter_list` only hold one dict?

File: python/paddle/optimizer/optimizer.py [hunk 10]
```diff
@@ -934,13 +1026,82 @@ def step(self):
...
+        Args:
+            param_group (dict): The group of Tensors to be optimzed with
+            different optimization options.
+        """
+        params = param_group['params']
+        if type(params) == Parameter:
+            param_group['params'] = [params]
+        elif isinstance(params, set):
+            raise TypeError(
+                "optimizer parameters should be in order collections,"
```
Comment (chenwhql) [inline python/paddle/optimizer/optimizer.py]: order -> ordered?

File: python/paddle/optimizer/optimizer.py (hunk 1 above)
Comment (jerrywgz) [reply python/paddle/optimizer/optimizer.py:100]: now the learning_rate in param groups is the scale of base learning rate, thus the lr of linear_2 is 0.1*0.1 = 0.01

File: python/paddle/optimizer/optimizer.py (hunk 2 above)
Comment (jerrywgz) [reply python/paddle/optimizer/optimizer.py:165]: Sure, param.weight_decay has the highest priority

File: python/paddle/optimizer/optimizer.py (hunk 3 above)
Comment (jerrywgz) [reply python/paddle/optimizer/optimizer.py]: done, thx

File: python/paddle/optimizer/optimizer.py (hunk 4 above)
Comment (jerrywgz) [reply python/paddle/optimizer/optimizer.py]: refined, this function will be rewrite in child class

File: python/paddle/optimizer/rmsprop.py (hunk 6 above)
Comment (jerrywgz) [reply python/paddle/optimizer/rmsprop.py]: sure, the default dict contains the extra argument

File: python/paddle/optimizer/optimizer.py (hunk 9 above)
Comment (jerrywgz) [reply python/paddle/optimizer/optimizer.py:187]: done, thx

File: python/paddle/optimizer/optimizer.py (hunk 7 above)
Comment (jerrywgz) [reply python/paddle/optimizer/optimizer.py]: done, thx

File: python/paddle/optimizer/optimizer.py (hunk 10 above)
Comment (jerrywgz) [reply python/paddle/optimizer/optimizer.py]: done, thx

**Review by zhiqiu — APPROVED** [review]
LGTM

**Review by ForFishes — APPROVED** [review]
LGTM

**Review by jzhang533 — APPROVED** [review]
lgtm
